### ISSUE TITLE
`TryFrom<&str> for KeyCode` for Deserialization

### ISSUE THREAD

**Issue by John2143** [bevyengine/bevy#957]
**What problem does this solve or what need does it fill?**

This will allow config parsers (ie. serde) to read user-given strings directly into bevy keycodes without having to copy the entire KeyCode enum into the game code.

**Describe the solution would you like?**

My current solution is to use a macro to map available keycodes into a list of [(&str, KeyCode]), then search on that.

https://github.com/John2143/bdo2/blob/e6a9a11d03db4c187db0ed07f3bdaf1e14d94f24/src/config_read.rs#L94-L120

The above version also provides the equivalent of a `impl<const N: usize> TryFrom<[&str; N]> for [KeyCode; N]`, and  `[&str; 1] -> KeyCode`, but that would be gated by `const-generics`. 

The solution of `TryFrom` would allow me to drop some of the serde code and remove the copy of KeyCode that I hold.

**Describe the alternative(s) you've considered?**

I'm not sure if this function exists already somewhere in the library, but I didn't see anything inside of keyboard.rs. I couldn't find any prior work by searching `bevy keycode from string`.

An alternative would be to provide a full blown `impl Deserialze for KeyCode`. This TryFrom is a requisite step that removes the most fragile part of my parsing right now. If the `Deserialize` impl is wanted, that could be a separate PR.

**Additional context**

https://github.com/John2143/bdo2/blob/e6a9a11d03db4c187db0ed07f3bdaf1e14d94f24/src/config_read.rs for a full implementation of Deserialize for KeyCode

Separate from #281 because that only adds `From<u32> for KeyCode`

May also be wanted on Mouse too, and perhaps a `&str` to `Option<Result<KeyCode, Mouse>>`

**Comment by amberkowalski** [comment]
Interesting thought, I see value in it.

**Comment by John2143** [comment]
The main question is if a `TryFrom<&str>` makes sense on KeyCode (and Mouse). I can't implement it on my own crate because of the orphan rule + lack of enum definition, so it needs to be added into the bevy definition of KeyCode.

I can submit a PR with the proposed addition if you'd like. It might be clearer than me writing it out.

**Comment by memoryruins** [comment]
If added, I think the impl should go through [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html). It would allow [`str::parse`](https://doc.rust-lang.org/std/primitive.str.html#method.parse) to be used without importing a trait.

**Comment by John2143** [comment]
I went to go try an implementation for it, but I found that #281, actually did solve what I was looking for. I thought that had only added the `repr` attribute.

The key is to add the optional serialize feature, which I couldn't find originally in the bevy docs. Serde basically provides the FromStr with that feature enabled.
